### ISSUE TITLE
Speckle Branching and Dev Practices 

### ISSUE THREAD

**Issue by fraguada** [speckleworks/SpeckleOrg#7]
Following up on the discussion on #semver slack channel and [semver issue raised on SpeckleRhino](https://github.com/speckleworks/SpeckleRhino/issues/97), we should discuss appropriate branching schemes for Speckle Repos. Here are some references shared by @pearswj and @didimitrie :

- https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow
- http://nvie.com/posts/a-successful-git-branching-model/
- http://scottchacon.com/2011/08/31/github-flow.html
- https://guides.github.com/introduction/flow/
- https://github.com/octokit/octokit.rb#versioning

**Comment by didimitrie** [comment]
I've changed the issue name to avoid confusion with the defunct Flux Flow thing 😬 

AFAIK we discussed with @pearswj (please correct me!): 

Taking the server repo as an example: new issues coming in will be judged by whether **breaking changes** are required or not and/or **tons of refactoring**, etc. If yes, they will go towards a `2.0.0+` milestone. If not, on, for example, a `1.1.+` milestone. 

Any dev work that goes into  `1.x.x` on the server happens in "hotfix"/"feature branches". Any work that goes towards a major version happens on its own feature branch, and gets merged into `dev`, which essentially is the next major branch version. 

Essentially, gitflow routines as described in the links posted by @fraguada  above apply for major version changes. For stuff that goes into `1.x.x` we have feature branches that go back into master. If the master moves ahead while deving your fix, you can always pull it again into your feature branch. 

Dang. Am I right?

**Comment by radumg** [comment]
spot-on :tada:  

actionable comments :
- this needs to be explained clearly in contributing section for people less familiar with git
- the issue & PR templates need to adjust for this (we might need `minor` & `major` templates)
- the proposal above to only maintain a dev branch for major releases is sound, but i suggest it be called `2.0` for example, making it clear it's not an indefinite, forever to be maintained & pushed to branch

supporting comments
- I think fully adopting the `git-flow` described in the `nvie` article is a bit overkill, especially for an OSS project with distributed contributors that should strive to make it easy & welcoming to contribute.
- git-flow is also essentially only using master as release tags, duplicating github functionality and by extension, this applies to maintaining a `dev` branch

**Comment by pearswj** [comment]
Can we clarify what the branching strategy is supposed to accomplish?

To start us off, my understanding is that Speckle in general is slightly different to the projects described by the git-flow blog post. Those projects only have one production instance, hence master describes the code that is currently running in production at any given time. Since anyone can run a Speckle server there will be (🤞) many servers and some people will inevitably get stuck on an old major version that may need the occasional bugfix release. Would you do this, or just say “tough luck, upgrade if you want the bug fixed”?

The other question is whether there needs to be a dedicated development branch for the next major version of Speckle. I’ve actually changed my mind on this one and would now argue that it might not be necessary since I think you’d be working to clear the deck for the previous major version before starting work (in any coordinated way) on any features slated for the next major version. Of course you can start work on these in feature branches and simply delay merging them into `develop` until after the last minor/patch release is cut.

Something to chew on...

**Comment by radumg** [comment]
@pearswj good stuff to chew on... 

I think you're right Speckle is different than what's described in `git-flow`, it's closer to NodeJS model, there'll be various versions being run out there in the wild. Node seem to have adopted something similar to what i described, where major versions are on a branch (`8.x` for example), which is retained for LTS and bugfixes, whilst development continues on `master` and `9.x`... unless i misunderstood them, their docs are also not great at explaining it 😛

**Comment by pearswj** [comment]
I had a good read through the links below and a look at some of their recent pull requests...

* https://github.com/nodejs/node/blob/master/COLLABORATOR_GUIDE.md#how-are-lts-branches-managed
* https://github.com/nodejs/Release

It looks like everything lands on `master` by default, and then gets nominated for backporting to an active/maintained release. Nominated changes land on the corresponding “staging” branch (e.g. `v9.x-staging`) then a release is cut by creating a “proposal” branch (e.g. `v9.1.1-proposal`) from staging the staging branch into the release branch (`v9.x`).

They seem to have a lot of processes, hence the vast amount of documentation (which still doesn’t cover everything!). There's some useful stuff there -- such as the distinction between development, staging and release branches -- but I'm not a fan of the selective backporting. I think the following workflow is much easier to keep track of and is closer to what's been happening organically so far...

1. Select the features that are going into the next version (i.e. `v1.1.0`)
2. Work on those features and any bug fixes and commit them to a development branch (`master`)
3. Once all the features are in, branch to create a staging branch (`release-v1.1.0`/`v1.1.0-staging`/etc.). Development for the next release/milestone can now continue in `master`
4. Test and refine release candidate (any new changes get merged back into development branch)
5. :shipit: and tag and/or merge into LTS branch (`v1.x`)

When something needs to be backported to a previous API version, it can most likely be handled on a case-by-case basis. The tags that Node use (i.e. `backport-requested-v1.x` and `backported-to-v1.x`) are nice for keeping track of this.

**Comment by radumg** [comment]
The only things i would argue slightly against in the above is that `master` is not guaranteed to be a working/stable branch at any point in time and that it requires manually keeping branches in sync (master & current release candidate).
Having that solves the recurring question `what's the currently stable & up-to-date branch ?`
Answer would be `Easy, master!` instead of `look up the latest release branch`

1.  Select the features that are going into the next version (i.e. v1.1.0)
2.  Work on those features and any bug fixes and commit them to a development branch `v1.1.0`
3.  Test and refine release candidate in this branch. Once all the features are in and accepted, merge to master 🎉   
4.  Development for the next release/milestone can now continue in `v1.2.0` or `2.0` branches (or whatever is next).
5.  :shipit: via Github releases. The original `v1.1.0` branch now becomes a LTS one, so bug-fixes can be cherry-picked back to it if necessary.

It's a minor variation but saves from maintaining 2 active branches and requiring people to commit to `master`.

**Comment by didimitrie** [comment]
Guys, this spiralled out of control a bit. I think the main point that we need to agree on now is how we push updates to the end-users on a distributed system - right? 

Bearing this needs to happen for:
- the server
- the web apps (could be a setInterval every day check on the server current version vs npm package latest version)
- the .net clients

I think we're adding a bit too much tape for a potentially much simpler  process? Only saying this as I do not expect speckle to go into the 10s or even 100s of prs a month any time soon! 

I don't want to discuss now LTSes, etc. The project overall is just not there yet, and missing a lot of functionality that still needs to go in 1.x.x. I can promise an LTS for the server & specs, but why should I for the viewer that needs a lot of work still, or the rhino client? 

Can we rewind and go back to: 

- you wanna fix/enhance something? branch off master and, most importantly, **write code**
  - is it less work? it's patch version bump
  - is it a lot of work? it's a minor version bump
  - does it need to break previous stuff? it's a major version bump, and you better have good reasons for it
- done? make a PR to master & grab a 🍺 

Releases: go up in minor versions, 1.1.0, 1.2.0, etc. and are composed of issues with the appropriate  milestone stuck to them

Release now = we scream at people to update on slack, etc.
Release in the future = we automatically notify people to update, nicely
